stdlib/public/core/Integers.swift

```diff
@@ -1501,14 +1501,20 @@ extension BinaryInteger {
   @inlinable
   @inline(__always)
   public func advanced(by n: Int) -> Self {
-    if Self.isSigned {
-      return self.bitWidth < n.bitWidth
-        ? Self(Int(truncatingIfNeeded: self) + n)
-        : self + Self(truncatingIfNeeded: n)
+    // This compares its maximum bit width versus Int.bitWidth.
+    // It should constant-fold since we use constant arguments.
+    let typeIsSmall  = Self.isSigned
+    ? Self(exactly:  Int.max) == nil
+    : Self(exactly: UInt.max) == nil
+    
+    if typeIsSmall {
+      return Self(Int(truncatingIfNeeded: self) + n)
```

Comment:
`FixedWidthInteger` imposes a requirement that the representable range be -2ⁿ ..< 2ⁿ, but `BinaryInteger` does not. So formally I think it's possible to have an unsigned `BinaryInteger` type `T` such that `T(exactly: UInt.max)` is nil, and yet `Int(truncatingIfNeeded: self)` can still wrap.

What do we lose if we define `typeIsSmall = Self(exactly: Int.max) == nil` unconditionally?

Reply:
Hm, I see. In that case, truncating `UInt(bitPattern: 0 &- n)` would be bad if  `Int.max` fits but `Int.min.magnitude` doesn't. The result would be negative, however, so trapping through `Self.init(_:)` would solve that. 

Alternatively, `typeIsSmall` could be defined as `Self(exactly: Int.min.magnitude) == nil`. If an unsigned integer takes the small integer path, it is at most `Int.max`. Otherwise, the unsigned integer can represent all `Int` magnitudes. `Int` would take the small integer path, but same-type conversions should turn into no-ops.

What do you think about `typeIsSmall: Bool = Self(exactly: Int.min.magnitude) == nil` instead?